src/python/pants/java/nailgun_executor.py

```diff
@@ -203,29 +208,54 @@ def __init__(self, timeout, stdout, stderr):
 
   def _await_socket(self, timeout):
     """Blocks for the nailgun subprocess to bind and emit a listening port in the nailgun stdout."""
-    with safe_open(self._ng_stdout, 'r') as ng_stdout:
-      start_time = time.time()
-      accumulated_stdout = ''
-      while 1:
-        # TODO: share the decreasing timeout logic here with NailgunProtocol.iter_chunks() by adding
-        # a method to pants.util.contextutil!
-        remaining_time = time.time() - (start_time + timeout)
-        if remaining_time > 0:
-          stderr = read_file(self._ng_stderr, binary_mode=True)
-          raise self.InitialNailgunConnectTimedOut(
-            timeout=timeout,
-            stdout=accumulated_stdout,
-            stderr=stderr,
-          )
-
-        readable, _, _ = select.select([ng_stdout], [], [], (-1 * remaining_time))
-        if readable:
-          line = ng_stdout.readline()                          # TODO: address deadlock risk here.
-          try:
-            return self._NG_PORT_REGEX.match(line).group(1)
-          except AttributeError:
-            pass
-          accumulated_stdout += line
+    start_time = time.time()
+    accumulated_stdout = ''
+
+    def calculate_remaining_time():
+      # TODO: share the decreasing timeout logic here with NailgunProtocol.iter_chunks() by adding
+      # a method to pants.util.contextutil!
+      return time.time() - (start_time + timeout)
+
+    def possibly_raise_timeout(remaining_time):
+      if remaining_time > 0:
+        stderr = read_file(self._ng_stderr, binary_mode=True)
+        raise self.InitialNailgunConnectTimedOut(
+          timeout=timeout,
+          stdout=accumulated_stdout,
+          stderr=stderr,
+        )
+
+    if PY3:
+      # NB: We use PollSelector, rather than the more efficient DefaultSelector, because
+      # DefaultSelector results in using the epoll() syscall on Linux, which does not work with
+      # regular text files like ng_stdout. See https://stackoverflow.com/a/8645770.
+      with selectors.PollSelector() as selector, \
+        safe_open(self._ng_stdout, 'r') as ng_stdout:
+        selector.register(ng_stdout, selectors.EVENT_READ)
+        while 1:
+          remaining_time = calculate_remaining_time()
+          possibly_raise_timeout(remaining_time)
+          events = selector.select(timeout=-1 * remaining_time)
```

Comment:
So we do this in a couple of places ("select until timeout, if readable do something"), I wonder if we could extract this to a common behaviour, which would hide away the `PY3` checks as well, something like:
```
def check_readable(file_descriptor: int, timeout: int, py3_selector=DefaultSelector):
  if PY3:
    # Use py3_selector for whatever
  else:
    # use select.select
```

Reply:
Good idea! I did this for two of the call-sites and it's a big improvement.

I didn't use the new function here, though, because I found it too difficult to work around this section using a while loop and registering the selector before that while loop. The control flow is too divergent for it to make sense rewiring `socket.is_readable()`.